clab/clab.go

```diff
@@ -113,6 +114,69 @@ func WithTopoFile(file, varsFile string) ClabOption {
 	}
 }
 
+// WithNodeFilter option sets a filter for nodes to be deployed.
+// A filter is a list of node names to be deployed,
+// names are provided exactly as they are listed in the topology file.
+// Since this is altering the clab.config.Topology.[Nodes,Links] it must only
+// be called after WithTopoFile.
+func WithNodeFilter(nodeFilter []string) ClabOption {
+	return func(c *CLab) error {
+		return filterClabNodes(c, nodeFilter)
+	}
+
+}
+
+func filterClabNodes(c *CLab, nodeFilter []string) error {
+	if len(nodeFilter) == 0 {
+		return nil
+	}
+
+	log.Infof("Applying node filter: %q", nodeFilter)
+
+	// newNodes := make(map[string]*types.NodeDefinition, len(c.Config.Topology.Nodes))
+	newLinks := make([]*types.LinkConfig, 0, len(c.Config.Topology.Links))
+
+	// filter nodes
+	for name := range c.Config.Topology.Nodes {
+		if exists := slices.Contains(nodeFilter, name); !exists {
+			log.Debugf("Excluding node %s", name)
+			delete(c.Config.Topology.Nodes, name)
+		}
+	}
+
+	// filter links
+	for _, l := range c.Config.Topology.Links {
+		// get the endpoints of the link and extract the node names
+		// to remove the links which have either side in the node filter
+		splitEpAside := strings.Split(l.Endpoints[0], ":")
+		if len(splitEpAside) != 2 {
+			continue
+		}
+
+		epA := splitEpAside[0]
+
+		splitEpBside := strings.Split(l.Endpoints[1], ":")
+		if len(splitEpBside) != 2 {
+			continue
+		}
+
+		epB := splitEpBside[0]
+
+		containsAside := slices.Contains(nodeFilter, epA)
+		containsBside := slices.Contains(nodeFilter, epB)
+
+		// if both endpoints of a link belong to the node filter, keep the link
+		if containsAside && containsBside {
+			log.Debugf("Including link %+v", l)
+			newLinks = append(newLinks, l)
+		}
+	}
+	// replace the original collection of links with the links that have both endpoints in the node filter
+	c.Config.Topology.Links = newLinks
+
+	return nil
+}
```

Comment:
Hi @steiler 
I changed this func a bit by removing the nested if and dealing with nodes map by deleting the elements from existing map. [ec0e1f9](https://github.com/srl-labs/containerlab/pull/1298/commits/ec0e1f99cd5f16da77e3e7d83f6ccd87963591f7)

I also carved out the function from the option func, so that we can unit test it and wrote a basic test.